### ISSUE TITLE
Mathematics of supervised/semi-supervised UMAP

### ISSUE THREAD

**Issue by renxwise** [lmcinnes/umap#135]
I notice UMAP can now work in supervised/semi-supervised manners. Where can I find the mathematics of these functions? I am curious how the data and labels are balanced.

**Comment by lmcinnes** [comment]
Sadly I don't have a formal write up of this yet. I'm having enough trouble trying to get the basic paper published, so I haven't been too concerned with formalising the extensions just yet.

The sketch is as follows: We can construct a fuzzy simplicial set (or 1-skeleton thereof; a weighted graph) given only a set of points and a metric defining distance between them. While one can do that for the actual data, one can also define a metric on the target values and generate a fuzzy simplicial set from that. Now we can view these two different metric spaces with the same points as two *views* of some underlying structure. Given that we can *intersect* the fuzzy simplicial sets, to arrive at the shared view. As an added step we will want to reassert the local-connectivity requirement (which is local to each point, so we have localised everything) and then recombine all the localised fuzzy simplicial sets by unioning them. The end result is a single fuzzy simplicial set that meets our desiderata and has the combined view of both the data and the target information.

Now, how does define a metric on target data? If the data is categorical you can essentially define a simple metric where points in the same category are distance zero, points in different categories are distance one, and if either point is uncategorized (for the semi-supervised case) it is distance 0.5. For other target types (say a regression problem) you can define other more sensible metrics. The code exposes this as the ``target_metric`` parameter. How does one balance the combination of data and labels? You can essentially apply a weighted intersection, or, alternatively, think of it as scaling the metrics relative to one another (so to give more weight to the categorical metric, choose a larger value than 1 for the distance between points in different categories). This is exposed in the code as ``target_weight``, which defaults to 0.5 (essentially balanced), and where 1.0 means use (essentially) only the target information, and 0.0 means use (essentially) only the data.

I know this is just a sketch, but hopefully it lays out the core ideas and gives you some idea of how you can play with things. All of this will hopefully be documented with some extra pages in the tutorial on advanced supervision techniques (including using the regression style supervision where the balance becomes more important).

**Comment by renxwise** [comment]
I have gotten the basic ideas. Thanks a lot.

**Comment by ummel** [comment]
Is UMAP capable of handling missing values (i.e. semi-supervised) for a numeric target type? Based on the description above, I sense that the answer should/could be "yes", but all of the examples I've seen so far consider a categorical target.

**Comment by ummel** [comment]
Yes, that makes sense. Thank you. What do you think about potential discretization of a continuous target (e.g. deciles) -- retaining the NA values (since they are allowed in the categorical case) -- and then using the custom `target_metric` approach to specify pairwise distances that better reflect the "natural" distance between categories in the underlying continuous data? I think the latter is related to the spirit of #140.

**Comment by lmcinnes** [comment]
I think that's an entirely reasonable approach bar one small issue. Right now categorical distance is handled separately because the number of possible distinct distances is (much) smaller than the number of distinct points. This means that "k nearest neighbors" is tricky because that potentially either includes points in the identical category (which may still number much larger than k), or essentially all points. That can be got around as long as you are going to intersect it with some other fuzzy simplicial set, and it's that trick that the code exploits right now. Trying to use a custom metric would send it to another code path, and that might have some issues if there are only a small number of distinct distance possible. You might alleviate some of that by avoiding the deciles as long as you can determine a suitable distance between NA values and actual values and put that in the custom metric.

All that to say: something along the lines of what you want is do-able, but there are some catches, so it might take some care and few attempts to get it right.

**Comment by lmcinnes** [comment]
We have not written a paper on the subject unfortunately -- there really isn't enough material for one. The idea is pretty straightforward (intersecting fuzzy simplicial sets / taking limits) and largely self-evident. The implementation is publicly available, and hopefully takes care of the technical matters of "making it work in practice". I would certainly welcome suggestions for where code comments might be added to make things clearer.
